src/irr/video/COpenGLPipelineCache.h

```diff
@@ -33,19 +34,31 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
 
 	void merge(uint32_t _count, const IGPUPipelineCache** _srcCaches) override
 	{
+		size_t sz = 0ull;
+		for (uint32_t i = 0u; i < _count; ++i)
+			sz = std::max(sz, static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache.getSize());
+
+		auto buf = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<decltype(m_cache)::MutablePairType>>(sz);
 		for (uint32_t i = 0u; i < _count; ++i)
 		{
 			const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache;
-			m_cache.insert(src.begin(), src.end());
+			src.outputAll(sz, buf->data());
+			for (size_t j = 0ull; j < sz; ++j)
+				m_cache.insert((*buf)[j].first, (*buf)[j].second);
```

Comment:
use a single-threaded cache, and mutex the shit out of the pipeline, just using a thread-safe cache won't make the member functions of the owner safe to call from multiple threads.

Reply:
> just using a thread-safe cache won't make the member functions of the owner safe to call from multiple threads.

why?

Reply:
hm, ok, i know what u mean, nvm